C#: Unify project name handling and fix issues with the handling of some special characters

RedworkDE opened the pull request:
- Resolves https://github.com/godotengine/godot/pull/74479#discussion_r1126752744
- Fixes https://github.com/godotengine/godot/issues/75619
- Instead of just adding a native version, I got rid of the C# version entirely and replaced it with a call to the native version.
- ~The native version should probably be moved into the mono module, but it is used in ProjectSettings for managing the C# feature to projects, so I left it there.~
- Also made the function directly take the assembly_name setting into account, so that it properly works everywhere.

RedworkDE committed:
C#: Unify project name handling and fix issues with the handling of some special characters

Co-authored-by: Raul Santos <raulsntos@gmail.com>

RedworkDE commented on core/config/project_settings.cpp:
```diff
@@ -68,12 +68,27 @@ String ProjectSettings::get_resource_path() const {
 	}
-	return safe_name;
+	name = get_setting_with_override("application/config/name");
+	Vector<String> invalid_chars = Vector<String>({ //
```
Empty comment exists to convince clang-format to let me keep the next comment where it is.

raulsntos commented on core/config/project_settings.cpp:
```diff
@@ -68,12 +68,27 @@ String ProjectSettings::get_resource_path() const {
-	String safe_name = OS::get_singleton()->get_safe_dir_name(get("application/config/name"));
-	if (safe_name.is_empty()) {
-		safe_name = "UnnamedProject";
+String ProjectSettings::get_csharp_project_name() const {
```
I don't know if the Core team will be happy with a C# related method in Project Settings. I think the `get_safe_project_name` method was meant to be agnostic since we are outside the .NET module.

And it seems to me that the only reason why this method is here is for the purpose of detecting C# usage for the project features. Maybe modules should be able to manipulate project features to customize/extend them.

raulsntos replied:
How about moving this method into the .NET module? Then ProjectSettings can check for the C# feature by looking for any .csproj file inside the project directory, instead of checking the exact path.

```diff
diff --git a/core/config/project_settings.cpp b/core/config/project_settings.cpp
index cb25dc9ebf..7e9d7dd98e 100644
--- a/core/config/project_settings.cpp
+++ b/core/config/project_settings.cpp
@@ -934,6 +934,23 @@ Error ProjectSettings::_save_custom_bnd(const String &p_file) { // add other par
 	return save_custom(p_file);
 }
 
+#ifdef TOOLS_ENABLED
+bool _csproj_exists(String p_root_dir) {
+	Ref<DirAccess> dir = DirAccess::open(p_root_dir);
+
+	dir->list_dir_begin();
+	String file_name = dir->_get_next();
+	while (file_name != "") {
+		if (!dir->current_is_dir() && file_name.get_extension() == "csproj") {
+			return true;
+		}
+		file_name = dir->_get_next();
+	}
+
+	return false;
+}
+#endif // TOOLS_ENABLED
+
 Error ProjectSettings::save_custom(const String &p_path, const CustomMap &p_custom, const Vector<String> &p_custom_features, bool p_merge_with_current) {
 	ERR_FAIL_COND_V_MSG(p_path.is_empty(), ERR_INVALID_PARAMETER, "Project settings save path cannot be empty.");
 
@@ -952,7 +969,7 @@ Error ProjectSettings::save_custom(const String &p_path, const CustomMap &p_cust
 		}
 	}
 	// Check for the existence of a csproj file.
-	if (FileAccess::exists(get_resource_path().path_join(get_safe_project_name() + ".csproj"))) {
+	if (_csproj_exists(get_resource_path())) {
 		// If there is a csproj file, add the C# feature if it doesn't already exist.
 		if (!project_features.has("C#")) {
 			project_features.append("C#");

```

RedworkDE commented:
Moved the `get_csharp_project_name` to the mono module as suggested, wasn't sure where to best put it so I added it to path_utils (suggestions welcome).

The project settings now check for any .csproj file as suggested.

raulsntos commented:
> Moved the `get_csharp_project_name` to the mono module as suggested, wasn't sure where to best put it so I added it to path_utils (suggestions welcome).

I think path_utils is fine, or maybe godotsharp_dirs.

akien-mga commented:
Thanks!
